File: unity/Assets/Script/Scene/Hall.cs
/**
 * Copyright 2024. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */


using UnityEngine;
using Com.Huawei.Game.Gobes;
using Com.Huawei.Game.Gobes.Utils;

public class Hall : MonoBehaviour
{
    public GameObject LoadingPre;

    public GameObject BackGroud;

    public Dialog Dailog;

    public bool Flag = true;

    public string Msg;

    // Update is called once per frame
    void Update()
    {
        if (!Flag)
        {
            Debug.Log("房间匹配失败");
            OpenFailWindow(Msg);
            Flag = true;
        }
        if (Global.room != null)
        {

            if (Global.isAsymmetric)
            {
                // 非对称匹配
                Route.GoAsymmetricRoom();
            }
            else
            {
                // 对称匹配
                Route.GoTeamRoom();
            }
        }
    }

    void Start()
    {
        InitListener();
    }

    public void InitListener()
    {
        Global.client.OnKickOff = () => showTips();
    }

    public void showTips()
    {
        Debug.Log("玩家被踢");
        UnityMainThread.wkr.AddJob(() => {
            OpenFailWindow("多端登录离线");
        });
    }
    //点击菜鸟区按钮
    public void OnOrdinaryRoomBtn()
    {
        Debug.Log("进入菜鸟区");
        Global.matchRule = "0";
        Route.GoMatch();
    }
    //点击高手区按钮
    public void OnExpertRoomBtn()
    {

        Debug.Log("进入高手区");
        Global.matchRule = "1";
        Route.GoMatch();
    }
    //点击快速匹配按钮
    public void OnFastMatchBtn()
    {
        Debug.Log("快速匹配");
        Debug.Log("弹出loading");
        GameObject loading = Instantiate(LoadingPre);
        GameObject bg = Instantiate(BackGroud);
        Debug.Log(Global.playerId);
        Debug.Log(Global.playerName);
        MatchPlayer();
    }

    void MatchPlayer() {

        MatchPlayerInfoParam matchPlayerInfoParam = new MatchPlayerInfoParam()
        {
            PlayerId = Global.playerId,
            MatchParams = Util.GetPlayerMatchParams()
        };

        // 需要根据是否非对称，选择对应的matchCode


        MatchTeamInfoParam matchTeamInfoParam = new MatchTeamInfoParam()
        {
            MatchParams = Util.getTeamMatchParams()
        };

        MatchPlayerConfig matchPlayerConfig = new MatchPlayerConfig()
        {
            MatchCode = Global.matchCode,
            PlayerInfo = matchPlayerInfoParam,
            TeamInfo = matchTeamInfoParam
        };
        PlayerConfig playerConfig = new PlayerConfig
        {
            CustomPlayerStatus = 0,
            CustomPlayerProperties = Util.getCustomPlayerProperties()
        };
        try
        {
            Global.client.MatchPlayer(matchPlayerConfig, playerConfig, response =>
            {
                if (response.RtnCode == 0)
                {
                    Debug.Log("MatchPlayer start");
                    Global.isOnlineMatch = true;
                    Util.SaveOnlineMatch(true);
                    Flag = true;
                }
                else
                {
                    Debug.Log("MatchPlayer failed");
                    Msg = "快速匹配失败"+Util.ErrorMessage(response);
                    Flag = false;
                }
                if (response.RtnCode == (int)ErrorCode.PLAYER_MATCH_CANCELED){
                    Msg = "快速匹配取消" + Util.ErrorMessage(response);
                }
            });
            Global.client.OnMatch = response =>
            {
                if (response.RtnCode == 0)
                {
                    Debug.Log("MatchPlayer success");
                    Global.Room = response.Room;
                    Global.player = response.Room._player;
                    Global.isOnlineMatch = true;
                    Util.SaveOnlineMatch(true);
                    Flag = true;
                }
                else
                {
                    Debug.Log("MatchPlayer failed");
                    Msg = "快速匹配失败"+Util.ErrorMessage(response);
                    Flag = false;
                }
                if (response.RtnCode == (int)ErrorCode.PLAYER_MATCH_CANCELED){
                    Msg = "快速匹配取消" + Util.ErrorMessage(response);
                }
            };
        }
        catch (SDKException e) {
            Msg = "快速匹配失败" + Util.ExceptionMessage(e);
            OpenFailWindow(Msg);
        }

    }

        //点击组队匹配按钮
        public void OnCreateTeamBtn()
    {
        if (!Util.IsInited())
        {
            OpenFailWindow("请先初始化 SDK");
            return;
        }
        CreateTeam();
    }

    private void CreateTeam() {
        Debug.Log("组队匹配");
        Debug.Log(Global.playerId);
        GameObject bg = Instantiate(BackGroud);
        CreateGroupConfig createGroupConfig = new CreateGroupConfig
        {
            MaxPlayers = 2,
            GroupName = "快乐小黑店",
            IsLock = 0,
            IsPersistent = 0,
            CustomPlayerStatus = "0",
            CustomPlayerProperties = Global.playerName,
            CustomGroupProperties = "快乐小黑店"
        };
        try
        {
            Global.group = Global.client.CreateGroup(createGroupConfig, CreateGroupCallback);
        }
        catch (SDKException e)
        {
            OpenFailWindow("创建小队失败" + Util.ExceptionMessage(e));
        }
    }

    public void CreateGroupCallback(CreateGroupBaseResponse res)   {
        if (res.RtnCode == 0)
        {
            Debug.Log("CreateGroup success");
            Global.group = res.Group;
            Route.GoTeam();
        }
        else
        {
            Debug.Log("CreateGroup failed");
            OpenFailWindow("创建小队失败" + Util.ErrorMessage(res));
        }

    }
    //点击加入队伍按钮
    public void OnJoinTeamBtn()
    {
        Debug.Log("加入队伍");
        Route.GoTeamInfoView();
    }

    public void OpenFailWindow(string FailMsg)
    {
        Debug.Log("弹出失败窗口");
        GameObject loading = GameObject.Find("/loading(Clone)");
        Destroy(loading);
        Dialog dialog = Instantiate(Dailog);
        dialog.Open("提示", FailMsg);
    }
}

File: cocos/assets/Script/scene/hall.ts
/**
 * Copyright 2024. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import global from "../../global";
import * as Util from "../../util";
import Dialog from "../comp/Dialog";
import Reloading from "../comp/Reloading";
import config from "../../config";
import {LockType} from "../commonValue";

const {ccclass, property} = cc._decorator;

@ccclass
export default class Hall extends cc.Component {

    @property(cc.Sprite)
    ordinaryRoomBtn: cc.Sprite = null;

    @property(cc.Sprite)
    expertRoomBtn: cc.Sprite = null;

    // 组队匹配
    @property(cc.Sprite)
    teamMatchBtn: cc.Sprite = null;

    // 快速匹配（可用）
    @property(cc.Sprite)
    fastMatchBtn: cc.Sprite = null;

    // 加入队伍
    @property(cc.Sprite)
    joinTeamBtn: cc.Sprite = null;

    // 战绩回放
    @property(cc.Sprite)
    recordBtn: cc.Sprite = null;

    // 正在匹配加载界面
    @property(cc.Prefab)
    reloadingPrefab: cc.Prefab = null;

    @property(cc.Prefab)
    dialogPrefab: cc.Prefab = null;

    private lockSubmit: boolean = false;

    start() {
        this.initView();
        this.initListener();
        this.initDialog();
    }

    initDialog() {
        // 设置对话框
        const dialogNode = cc.instantiate(this.dialogPrefab) as cc.Node;
        dialogNode.parent = this.node;
    }

    /**
     * 界面初始化
     */
    initView() {
        // 设置加载Dialog
        const reloadingNode = cc.instantiate(this.reloadingPrefab) as cc.Node;
        reloadingNode.parent = this.node;
    }

    initListener() {
        this.ordinaryRoomBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onOrdinaryRoomBtn());
        this.expertRoomBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onExpertRoomBtn());
        this.teamMatchBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onTeamMatchBtn());
        this.joinTeamBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onJoinTeamBtn());
        this.fastMatchBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onMatchPlayer());
        this.recordBtn.node.on(cc.Node.EventType.TOUCH_START, () => this.onRecordList());

        // 绑定在线/组队匹配监听事件
        global.client.onMatch.clear();
        global.client.onMatch((onMatchResponse) => this.onMatch(onMatchResponse));
        global.client.onKickOff(()=>this.onKickOff())
    }

    /**
     * “战绩回放”按钮点击事件
     */
    onRecordList() {
        cc.director.loadScene('recordList');
        /*global.client.queryRecordList(1, 8).then((res) => {
            console.log('queryRecordList success res: ' + JSON.stringify(res));
        }).catch((err) => {
            console.log('queryRecordList err: ' + err);
        })*/
    }

    /**
     * “普通房间”按钮点击事件
     */
    onOrdinaryRoomBtn() {
        Util.printLog(`正在进入菜鸟区`);
        global.matchRule = '0';
        cc.director.loadScene("match");
    }

    /**
     * “高手房间”按钮点击事件
     */
    onExpertRoomBtn() {
        Util.printLog(`正在进入高手区`);
        global.matchRule = '1';
        cc.director.loadScene("match");
    }

    onTeamMatchBtn() {
        if (!Util.isInited()) {
            return Util.printLog("请先初始化 SDK");
        }
        !this.lockSubmit && this.teamMatch();
    }

    /**
     * 组队匹配（创建队伍）
     */
    teamMatch() {
        let data = {
            playerName: global.playerName
        };
        global.client.createGroup({
            maxPlayers: 2,
            groupName: "快乐小黑店",
            customGroupProperties: "",
            isLock: LockType.UnLocked,    // 是否禁止加入 0:不禁止 1:禁止 默认0
            isPersistent: 0,    // 是否持久化 0:不持久化 1:持久化 默认0
        }, {
            customPlayerStatus: 0,
            customPlayerProperties: JSON.stringify(data),
        }).then((group) => {
            Util.printLog("队伍创建成功");
            global.group = group;
            this.lockSubmit = false;
            cc.director.loadScene("team");
        }).catch((e) => {
            this.lockSubmit = false;
            Dialog.open("提示", "组队匹配失败" + Util.errorMessage(e));
        });
    }

    /**
     * 加入队伍
     */
    onJoinTeamBtn() {
        cc.director.loadScene("teaminfo");
    }

    onDisable() {
        // 关闭对话框
        Dialog.close();
    }

    /**
     * “快速匹配”按钮
     * 进行在线匹配
     */
    async onMatchPlayer() {
        Util.printLog("玩家 " + global.playerId + " 进行在线匹配");
        let player = {
            playerId: global.playerId,
            matchParams: Util.getPlayerMatchParams()
        };
        this.lockSubmit = true;
        Reloading.open("匹配中。。。", true, () => {
            this.cancelTeamMatch();
        });
        // 调用GOBE的matchPlayer发起在线匹配
        global.client.matchPlayer(
            {
                playerInfo: player,
                teamInfo: Util.getTeamMatchParams(),
                matchCode: config.asymmetric ? config.asymmetricMatchCode : config.matchCode
            }, {customPlayerStatus: 0, customPlayerProperties: Util.getCustomPlayerProperties()}).then((res) => {
            // 匹配开始
            global.isOnlineMatch = true;
            Util.printLog("在线匹配开始")
        }).catch((e) => {
            this.lockSubmit = false;
            Util.printLog("在线匹配失败" + Util.errorMessage(e));
            Util.printLog(e.code);
            Reloading.close();
        });
    }

    onMatch(res){
        if (res.rtnCode === 0) {
            Util.printLog('在线匹配成功:' + res.room);
            global.room = res.room;
            global.player = res.room.player;
            cc.director.loadScene(config.asymmetric ? "asymmetricroom" : "teamroom");
            Reloading.close();
        } else {
            this.lockSubmit = false;
            Util.printLog("在线匹配失败" + Util.errorMessage(res));
            Reloading.close();
        }
    }

    /**
     * 取消快速匹配
     */
    cancelTeamMatch() {
        global.client.cancelMatch().then(() => {
            global.isOnlineMatch = false;
            Util.printLog('取消匹配成功');
        }).catch(() => {
            Util.printLog('取消匹配失败');
        })
    }

    private onKickOff() {
        Util.printLog('多端登录，玩家离线');
    }
}
